File: scripts/resolve-maps.py
#!/usr/bin/env python3
"""
resolve-maps.py
---------------
Scans data/projects/completed.js and ongoing.js for any project that has
a mapsUrl but no mapsEmbed. Resolves the short link to coordinates and
injects a coordinate-based mapsEmbed so the embedded map shows a precise pin.

For projects that have mapsQuery but no mapsUrl/mapsEmbed, falls back to
geocoding via Nominatim (OpenStreetMap) to inject a coordinate-based embed.

Usage:
    python3 scripts/resolve-maps.py

Run this once after adding a new project with a mapsUrl. The data files
are updated in place — no other code changes needed.
"""

import re
import sys
import json
import time
import urllib.request
import urllib.parse

DATA_FILES = [
    'data/projects/completed.js',
    'data/projects/ongoing.js',
]

def resolve_url(short_url):
    """Follow redirects and return the final URL."""
    try:
        req = urllib.request.Request(
            short_url,
            headers={'User-Agent': 'Mozilla/5.0'}
        )
        res = urllib.request.urlopen(req, timeout=10)
        return res.url
    except Exception as e:
        print(f'  ERROR resolving {short_url}: {e}')
        return None

def extract_coords(full_url):
    """Extract lat,lng from a resolved Google Maps URL."""
    # Format: /@lat,lng,zoom  or  /place/lat,lng/
    match = re.search(r'/@(-?\d+\.\d+),(-?\d+\.\d+)', full_url)
    if match:
        return match.group(1), match.group(2)
    # Fallback: plain coordinates in path
    match = re.search(r'/place/(-?\d+\.\d+),(-?\d+\.\d+)', full_url)
    if match:
        return match.group(1), match.group(2)
    # Fallback: q=lat,lng in query string
    match = re.search(r'[?&]q=(-?\d+\.\d+),(-?\d+\.\d+)', full_url)
    if match:
        return match.group(1), match.group(2)
    return None, None

def geocode_query(maps_query):
    """Geocode a mapsQuery string via Nominatim and return (lat, lng) or (None, None)."""
    # Convert URL-encoded plus-separated query to plain text
    address = maps_query.replace('+', ' ')
    params = urllib.parse.urlencode({'q': address, 'format': 'json', 'limit': '1'})
    url = f'https://nominatim.openstreetmap.org/search?{params}'
    try:
        req = urllib.request.Request(
            url,
            headers={'User-Agent': 'resolve-maps/1.0 (construction-website)'}
        )
        res = urllib.request.urlopen(req, timeout=10)
        data = json.loads(res.read().decode())
        if data:
            return data[0]['lat'], data[0]['lon']
    except Exception as e:
        print(f'  ERROR geocoding "{address}": {e}')
    return None, None

def build_embed_url(lat, lng, zoom=17):
    return f'https://maps.google.com/maps?q={lat},{lng}&z={zoom}&output=embed'

def process_file(filepath):
    with open(filepath, 'r') as f:
        content = f.read()

    original = content

    # ── Pass 1: resolve mapsUrl entries that have no mapsEmbed ──
    pattern = re.compile(
        r'(mapsUrl:\s*"(https://maps\.app\.goo\.gl/[^"]+)",'
        r'\s*\n(?!\s*mapsEmbed:))'
    )

    matches = list(pattern.finditer(content))
    if matches:
        offset = 0
        for m in matches:
            url = m.group(2)
            print(f'  Resolving mapsUrl: {url}')
            full_url = resolve_url(url)
            if not full_url:
                continue
            lat, lng = extract_coords(full_url)
            if not lat:
                print(f'    Could not extract coordinates from: {full_url}')
                continue
            print(f'    Coordinates: {lat}, {lng}')
            embed_url = build_embed_url(lat, lng)
            insert_after = m.end() + offset
            insert_text = f'    mapsEmbed: "{embed_url}",\n'
            content = content[:insert_after] + insert_text + content[insert_after:]
            offset += len(insert_text)
    else:
        print(f'  {filepath}: no mapsUrl entries need resolving.')

    # ── Pass 2: geocode mapsQuery entries that still have no mapsEmbed ──
    # Find all project blocks: look for mapsQuery lines not preceded by mapsEmbed anywhere in that project
    # Strategy: find each mapsQuery line, then check if mapsEmbed exists nearby (within ~10 lines)
    lines = content.split('\n')
    new_lines = []
    i = 0
    changed_in_pass2 = False
    while i < len(lines):
        line = lines[i]
        mq_match = re.match(r'(\s*)mapsQuery:\s*"([^"]+)",', line)
        if mq_match:
            indent = mq_match.group(1)
            maps_query = mq_match.group(2)
            # Check if mapsEmbed already exists within the next 15 lines
            window = lines[i+1:i+16]
            has_embed = any('mapsEmbed:' in l for l in window)
            # Also check the previous 15 lines
            prev_window = lines[max(0, i-15):i]
            has_embed = has_embed or any('mapsEmbed:' in l for l in prev_window)

            new_lines.append(line)
            if not has_embed:
                print(f'  Geocoding mapsQuery: {maps_query}')
                lat, lng = geocode_query(maps_query)
                time.sleep(1)  # Nominatim rate limit: 1 req/sec
                if lat and lng:
                    print(f'    Coordinates: {lat}, {lng}')
                    embed_url = build_embed_url(lat, lng)
                    new_lines.append(f'{indent}mapsEmbed: "{embed_url}",')
                    changed_in_pass2 = True
                else:
                    print(f'    Could not geocode: {maps_query}')
        else:
            new_lines.append(line)
        i += 1

    if changed_in_pass2:
        content = '\n'.join(new_lines)

    if content != original:
        with open(filepath, 'w') as f:
            f.write(content)
        print(f'  Updated: {filepath}')
    else:
        print(f'  No changes: {filepath}')

if __name__ == '__main__':
    print('Resolving map URLs...\n')
    for f in DATA_FILES:
        print(f'Processing {f}')
        process_file(f)
    print('\nDone. Reload the page to see updated maps.')

File: scripts/resolve-maps.js
#!/usr/bin/env node
/**
 * resolve-maps.js
 * ---------------
 * Reads all project data files and resolves a Google Maps embed URL
 * for every project, using whichever source is available:
 *
 *   1. mapsUrl  (maps.app.goo.gl short link) → follow redirect → extract lat/lng
 *   2. mapsQuery (address string)             → Nominatim geocode → lat/lng
 *
 * Writes: data/projects/maps.js
 *   const MAPS_EMBEDS = { "project-id": "https://maps.google.com/maps?q=lat,lng&z=17&output=embed", ... }
 *
 * Usage:
 *   node scripts/resolve-maps.js
 *
 * Run once after adding any new project. Re-run to refresh all pins.
 * The data files (completed.js / ongoing.js) do NOT need mapsEmbed — this
 * script generates the embed URLs automatically.
 */

const https = require('https');
const http  = require('http');
const fs    = require('fs');
const path  = require('path');
const url   = require('url');

// ── Config ──────────────────────────────────────────────────────────────────

const DATA_FILES = [
  'data/projects/completed.js',
  'data/projects/ongoing.js',
];
const OUT_FILE  = 'data/projects/maps.js';
const EMBED_ZOOM = 17;

// ── Helpers ──────────────────────────────────────────────────────────────────

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/** Follow all redirects and return the final URL. */
function resolveRedirect(shortUrl) {
  return new Promise((resolve) => {
    const parsed = url.parse(shortUrl);
    const lib = parsed.protocol === 'https:' ? https : http;
    const req = lib.request(
      { hostname: parsed.hostname, path: parsed.path, method: 'HEAD',
        headers: { 'User-Agent': 'Mozilla/5.0' } },
      (res) => {
        if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
          resolveRedirect(res.headers.location).then(resolve);
        } else {
          resolve(shortUrl); // no more redirects — use as-is (GET fallback)
        }
      }
    );
    req.on('error', () => resolve(null));
    req.end();
  });
}

/** Follow redirects via GET (needed when HEAD doesn't expose final URL). */
function resolveRedirectGet(shortUrl) {
  return new Promise((resolve) => {
    const parsed = url.parse(shortUrl);
    const lib = parsed.protocol === 'https:' ? https : http;
    const req = lib.request(
      { hostname: parsed.hostname, path: parsed.path, method: 'GET',
        headers: { 'User-Agent': 'Mozilla/5.0' } },
      (res) => {
        if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
          resolve(res.headers.location);
        } else {
          resolve(shortUrl);
        }
        res.destroy();
      }
    );
    req.on('error', () => resolve(null));
    req.end();
  });
}

/** Extract lat,lng from a resolved Google Maps URL. */
function extractCoords(fullUrl) {
  // /@lat,lng,zoom
  let m = fullUrl.match(/@(-?\d+\.\d+),(-?\d+\.\d+)/);
  if (m) return [m[1], m[2]];
  // /place/lat,lng
  m = fullUrl.match(/\/place\/(-?\d+\.\d+),(-?\d+\.\d+)/);
  if (m) return [m[1], m[2]];
  // ?q=lat,lng
  m = fullUrl.match(/[?&]q=(-?\d+\.\d+),(-?\d+\.\d+)/);
  if (m) return [m[1], m[2]];
  return null;
}

/** Geocode an address string via Nominatim. Returns [lat, lng] or null. */
function geocode(address) {
  return new Promise((resolve) => {
    const query = encodeURIComponent(address + ', India');
    const reqUrl = `https://nominatim.openstreetmap.org/search?q=${query}&format=json&limit=1`;
    https.get(reqUrl, { headers: { 'User-Agent': 'resolve-maps/1.0 (mm-constructions-website)' } }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const results = JSON.parse(data);
          if (results.length) resolve([results[0].lat, results[0].lon]);
          else resolve(null);
        } catch { resolve(null); }
      });
    }).on('error', () => resolve(null));
  });
}

function buildEmbedUrl(lat, lng, zoom = EMBED_ZOOM) {
  return `https://maps.google.com/maps?q=${lat},${lng}&z=${zoom}&output=embed`;
}

/** Parse project objects from a JS data file using regex (no eval). */
function extractProjects(content) {
  const projects = [];
  // Match each { ... } object block at the top level of the array
  const blockRe = /\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\}/g;
  let m;
  while ((m = blockRe.exec(content)) !== null) {
    const block = m[0];
    const getField = (key) => {
      const r = new RegExp(`${key}:\\s*"([^"]+)"`);
      const fm = block.match(r);
      return fm ? fm[1] : null;
    };
    const id = getField('id');
    if (!id) continue;
    projects.push({
      id,
      mapsUrl:   getField('mapsUrl'),
      mapsQuery: getField('mapsQuery'),
    });
  }
  return projects;
}

// ── Main ─────────────────────────────────────────────────────────────────────

async function resolveProject(p) {
  // Strategy 1: short URL redirect
  if (p.mapsUrl) {
    console.log(`  [${p.id}] Resolving mapsUrl: ${p.mapsUrl}`);
    let finalUrl = await resolveRedirectGet(p.mapsUrl);
    if (finalUrl) {
      // May need another hop
      if (finalUrl.includes('maps.app.goo.gl') || finalUrl.includes('goo.gl')) {
        finalUrl = await resolveRedirectGet(finalUrl);
      }
      const coords = finalUrl ? extractCoords(finalUrl) : null;
      if (coords) {
        console.log(`    → ${coords[0]}, ${coords[1]}`);
        return buildEmbedUrl(coords[0], coords[1]);
      }
      console.log(`    Redirect resolved to: ${finalUrl}`);
      console.log(`    Could not extract coordinates — falling back to geocoding`);
    }
  }

  // Strategy 2: geocode the address
  if (p.mapsQuery) {
    const address = p.mapsQuery.replace(/\+/g, ' ');
    console.log(`  [${p.id}] Geocoding: ${address}`);
    await sleep(1100); // Nominatim rate limit: 1 req/sec
    const coords = await geocode(address);
    if (coords) {
      console.log(`    → ${coords[0]}, ${coords[1]}`);
      return buildEmbedUrl(coords[0], coords[1]);
    }
    // Try with shortened city name as fallback
    const parts = address.split(' ');
    const fallback = parts.slice(Math.max(0, parts.length - 3)).join(' ');
    if (fallback !== address) {
      console.log(`    Retrying with: ${fallback}`);
      await sleep(1100);
      const coords2 = await geocode(fallback);
      if (coords2) {
        console.log(`    → ${coords2[0]}, ${coords2[1]}`);
        return buildEmbedUrl(coords2[0], coords2[1]);
      }
    }
    console.log(`    Could not geocode`);
  }

  return null;
}

async function main() {
  const embeds = {};

  for (const file of DATA_FILES) {
    console.log(`\nProcessing ${file}`);
    const content = fs.readFileSync(file, 'utf8');
    const projects = extractProjects(content);
    console.log(`  Found ${projects.length} projects`);

    for (const p of projects) {
      const embedUrl = await resolveProject(p);
      if (embedUrl) embeds[p.id] = embedUrl;
    }
  }

  // Write output file
  const lines = Object.entries(embeds).map(
    ([id, u]) => `  "${id}": "${u}"`
  );
  const output = [
    '// Auto-generated by scripts/resolve-maps.js — do not edit manually.',
    '// Run: node scripts/resolve-maps.js',
    `// Generated: ${new Date().toISOString()}`,
    'const MAPS_EMBEDS = {',
    lines.join(',\n'),
    '};',
    '',
  ].join('\n');

  fs.writeFileSync(OUT_FILE, output);
  console.log(`\nWrote ${Object.keys(embeds).length} embed URLs to ${OUT_FILE}`);
  console.log('Done. Reload the page to see updated maps.');
}

main().catch(err => { console.error(err); process.exit(1); });
